[Memory Snapshot] Add recordAnnotations to capture record_function annotations

aaronenyeshi opened the pull request:
Summary: Add new traceEvents into Memory Snapshot for record_function annotations. These will capture both the profiler's step annotation as well as user annotations.

Test Plan:
CI

New Snapshot Generated:
devvm2184.cco0.facebook.com.Apr_19_13_27_14.3072800.snapshot.pickle

Snippet of Snapshot device_traces show `ProfilerStep#0`, and `## forward ##` annotations:
```
[[{'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168556,
   'frames': [{'name': 'START', 'filename': 'ProfilerStep#0', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168738,
   'frames': [{'name': 'END', 'filename': 'ProfilerStep#0', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168865,
   'frames': [{'name': 'START', 'filename': 'ProfilerStep#1', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168920,
   'frames': [{'name': 'START', 'filename': '## forward ##', 'line': 0}]},
  {'action': 'alloc',
   'addr': 140166073581568,
   'size': 3211264,
   'stream': 0,
   'time_us': 1713558427172978,
   'frames': [{'name': '_conv_forward',
     'filename': '/mnt/xarfuse/uid-416185/235d4caf-seed-nspid4026531836_cgpid32884718-ns-4026531840/torch/nn/modules/conv
```

Differential Revision: D55941362

Pulled By: aaronenyeshi




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -170,8 +170,10 @@ struct TraceEntry {
-        // bytes reported by cuda)
+    OOM, // the allocator threw an OutOfMemoryError (addr_ is the amount of free
+         // bytes reported by cuda)
+    ANNOTATION_START, // API call made to record the start of an annotation
```
I'd like to avoid the lists of actions here growing to add additional event types. Can we just have a single 'USER_DEFINED' event here that we use for both annotation start/end, and then put the additional information into gathered context?

aaronenyeshi replied:
Changed to USER_DEFINED

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -180,13 +182,15 @@ struct TraceEntry {
       approx_time_t time,
-      std::shared_ptr<GatheredContext> context = nullptr)
+      std::shared_ptr<GatheredContext> context = nullptr,
+      std::string name = "")
```
I'd also like to avoid adding new fields to this struct that only are used for custom annotations. Is it possible to put the name into the GatheredContext object, which already holds information about "where" this event occurred For instance, we can add an additional "frame" to the top of the stack trace. That would make these events easier to handle in viz as well, since the data would get threaded into the frames object. Otherwise there will be a ton of places that would have to be modified to get the name field threaded through to the right parts to visualize.

aaronenyeshi replied:
This is a great idea! I've updated the PR to save the name and stage (START or END) into a frame in the GatheredContext. This makes sense to me! Thanks!

zdevito commented on torch/csrc/autograd/record_function_ops.cpp:
```diff
@@ -30,6 +31,8 @@ static void record_function_enter(
       rec.before(name);
     }
   }
+  // Record the annotation into CCA for memory snapshot tool.
```
`rec.before` seems like an existng facility for doing callbacks when a recording starts that this should probably be using. Then the actual thing can be registered from a place that can import the cuda header, and we can avoid having to put functions in all the other allocator.s

aaronenyeshi replied:
Great point! actually that saves use a lot of hacky static function callbacks, this facility seems much easier to read and maintain. Thanks!

zdevito reviewed (commented):
This will help get the visualization linked to custom events.  I have some inline suggestions about how to minimize the number of places that will need to change.

aaronenyeshi commented:
Addressed Zach's comments, removed the name field from TraceEntry, instead we can embed the record_function annotation name and stage into the frames of the user_defined TraceEntry. Replaced annotation_start and annotation_end with single user_defined. Used record_function.h callbacks to register the CCA record annotation function. Also, tried to fix all lints from CLANGTIDY, CLANGFORMAT, and unquoted includes.

zdevito commented on torch/csrc/profiler/combined_traceback.h:
```diff
@@ -58,10 +58,17 @@ struct TORCH_API CapturedTraceback : public c10::GatheredContext {
   int traversePython(visitproc visit, void* arg);
   int clearPython();
 
+  void record_user_defined_frame(
```
nit: can we match the method style for this class, `recordUserDefinedFrame`

zdevito replied:
Let's have this API take a Frame object directly rather than name/stage so that it doesn't become specific to record function.

zdevito reviewed (approved):
Looks good! Thanks for making those changes.

aaronenyeshi commented:
Weird that test_tensorexpr is getting SIGSEGV for this PR. 
https://hud.pytorch.org/pr/pytorch/pytorch/124179#24080265463

aaronenyeshi commented:
Weird, I cannot reproduce the crashing test_tensorexpr.py locally:
```
$ python test_tensorexpr.py
/home/aaronshi/.conda/envs/pytorch/lib/python3.11/site-packages/torch/_custom_ops.py:253: DeprecationWarning: torch.library.impl_abstract was renamed to torch.library.register_fake. Please use that instead; we will remove torch.library.impl_abstract in a future version of PyTorch.
  return torch.library.impl_abstract(qualname, func, _stacklevel=2)
................s........s......s.................................s.......
----------------------------------------------------------------------
Ran 74 tests in 27.040s

OK (skipped=4)
```

aaronenyeshi commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "test_tensorexpr.py::TestTensorExprFuser::test_simple_add is causing a segfault https://github.com/pytorch/pytorch/actions/runs/9097383783/job/25007155440 https://hud.pytorch.org/pytorch/pytorch/commit/187aeaeabf612824c2d0e9be72f80ce6612760d4, test was skipped due to bad TD" -c nosignal

aaronenyeshi commented:
@pytorchbot rebase -b main

aaronenyeshi commented:
@pytorchbot rebase -b main

aaronenyeshi commented:
Failures seem unrelated. In the past, this PR was passing everything except tensorexpr, which david helped to fix.

[Docathon Labels Sync / check-labels](https://hud.pytorch.org/pr/pytorch/pytorch/124179#26386155684) ([gh](https://github.com/pytorch/pytorch/actions/runs/9570727423/job/26386155684))
[inductor / cuda12.1-py3.10-gcc9-sm86 / test (inductor, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/124179#26388100263) ([gh](https://github.com/pytorch/pytorch/actions/runs/9570729287/job/26388100263))
[inductor / cuda12.1-py3.12-gcc9-sm86 / test (inductor, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/124179#26388656911) ([gh](https://github.com/pytorch/pytorch/actions/runs/9570729287/job/26388656911))

Docathon fails because of regex check on ProfilerStep#0, and inductor failures look to be reproducible on other PRs too.

aaronenyeshi commented:
@pytorchbot merge -i

aaronenyeshi committed:
[Memory Snapshot] Add recordAnnotations to capture record_function annotations (#124179)

Summary:
Add new traceEvents into Memory Snapshot for record_function annotations. These will capture both the profiler's step annotation as well as user annotations.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/124179

Test Plan:
CI

New Snapshot Generated:
devvm2184.cco0.facebook.com.Apr_19_13_27_14.3072800.snapshot.pickle

Snippet of Snapshot device_traces show `ProfilerStep#0`, and `## forward ##` annotations:
```
[[{'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168556,
   'frames': [{'name': 'START', 'filename': 'ProfilerStep#0', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168738,
   'frames': [{'name': 'END', 'filename': 'ProfilerStep#0', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168865,
   'frames': [{'name': 'START', 'filename': 'ProfilerStep#1', 'line': 0}]},
  {'action': 'user_defined',
   'addr': 0,
   'size': 0,
   'stream': 0,
   'time_us': 1713558427168920,
   'frames': [{'name': 'START', 'filename': '## forward ##', 'line': 0}]},
  {'action': 'alloc',
   'addr': 140166073581568,
   'size': 3211264,
   'stream': 0,
   'time_us': 1713558427172978,
   'frames': [{'name': '_conv_forward',
     'filename': '/mnt/xarfuse/uid-416185/235d4caf-seed-nspid4026531836_cgpid32884718-ns-4026531840/torch/nn/modules/conv
```

Pulled By:
aaronenyeshi

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

aaronenyeshi

Differential Revision: D55941362

aaronenyeshi commented:
Looks like this PR need to be reopened since it was reverted before.
